e2e_tests/E2ETests.py

```diff
@@ -299,43 +303,98 @@ def run_auto_submit(self, timeout: int = 10, sleep_time: int = 10):
             p = exe(cmd, debug=False, non_blocking=True, stdout=out_f, stderr=err_f)
 
-        # Monitor mgmt db
-        print("Mgmt DB progress: ")
-        timeout = timeout * 60
-        start_time = time.time()
-        while time.time() - start_time < timeout:
-            try:
-                total_count, comp_count, failed_count = self.check_mgmt_db_progress()
-            except sql.OperationalError as ex:
+            # Monitor mgmt db
+            print("Mgmt DB progress: ")
+            timeout = timeout * 60
+            start_time = time.time()
+            while time.time() - start_time < timeout:
+                try:
+                    total_count, comp_count, failed_count = (
+                        self.check_mgmt_db_progress()
+                    )
+                except sql.OperationalError as ex:
+                    print(
+                        "Operational error while accessing database. "
+                        "Retrying in {} seconds\n{}".format(sleep_time, ex)
+                    )
+                    continue
+
                 print(
-                    "Operational error while accessing database. "
-                    "Retrying in {} seconds\n{}".format(sleep_time, ex)
+                    "Completed: {}, Failed: {}, Total: {}".format(
+                        comp_count, failed_count, total_count
+                    )
                 )
-                continue
 
-            print(
-                "Completed: {}, Failed: {}, Total: {}".format(
-                    comp_count, failed_count, total_count
+                if proc_type_cancel:
+                    proc_type_cancel = self.cancel_running(proc_type_cancel)
+
+                if total_count == (comp_count + failed_count):
+                    break
+                else:
+                    time.sleep(sleep_time)
+
+            if time.time() - start_time > timeout:
+                self.errors.append(
+                    Error("Auto-submit timeout", "The auto-submit timeout expired.")
                 )
-            )
+        # Still display the exception
+        except Exception as ex:
+            raise ex
+        # Clean up
+        finally:
+            if p is not None:
+                p.terminate()
+            if out_f is not None:
+                out_f.close()
+            if err_f is not None:
+                err_f.close()
+
+    def cancel_running(self, proc_types: List[const.ProcessType]):
+        """Looks for any running task of the specified process types
+        and attempts to cancel one of each.
+        """
+        # Get all running jobs in the mgmt db
+        proc_types_int = [proc_type.value for proc_type in proc_types]
+        with connect_db_ctx(sim_struct.get_mgmt_db(self.stage_dir)) as cur:
+            entries = cur.execute(
+                "SELECT proc_type, job_id  FROM state "
+                "WHERE status == 3 AND proc_type IN ("
+                + "?," * (len(proc_types_int) - 1)
```

Comment:
Arguably nicer to use python slicing.

Reply:
Could use "IN ({}?)".format("?,"*(len(proc_types_int) - 1)) ?

Reply:
That's nicer still. Don't have to do string appending this way.